src/jet_client.rs

```diff
@@ -144,224 +151,331 @@ impl Future for JetMsgReader {
+                instance: self.config.jet_instance().clone(),
+                timeout: ACCEPT_REQUEST_TIMEOUT_SEC,
+            });
+            let mut response_msg_buffer = Vec::with_capacity(512);
+            response_msg.write_to(&mut response_msg_buffer)?;
+
+            Ok(Async::Ready(response_msg_buffer))
+        } else {
+            Ok(Async::NotReady)
         }
+    }
 
-        // We have a response for sure ==> Send response
-        let response_msg = self.response_msg.as_ref().expect("We must have a response to send");
-        let mut v = Vec::new();
-        response_msg.write_to(&mut v)?;
-        try_ready!(self.transport.poll_write(&v));
-
-        // Start timeout to remove the association if no connect is received. We start it only if a new association has just been added,
-        // possible only with version 1.
-        if let JetMessage::JetAcceptRsp(accept_rsp) = response_msg {
-            if accept_rsp.version == 1 && accept_rsp.status_code == StatusCode::OK {
-                start_remove_association_future(self.executor_handle.clone(), self.jet_associations.clone(), accept_rsp.association);
+    fn handle_set_transport(&mut self) -> Poll<(), io::Error> {
+        if let Ok(mut jet_associations) = self.jet_associations.try_lock() {
+            match self.request_msg.version {
+                1 => {
+                    let association = jet_associations
+                        .get_mut(
+                            self.association_uuid
+                                .as_ref()
+                                .expect("Must be set during parsing of the request"),
+                        )
+                        .expect("Was checked during parsing the request");
+                    let candidate = association
+                        .get_candidate_by_index(0)
+                        .expect("Only one candidate exists in version 1 and there is no candidate id");
+                    candidate.set_transport(self.transport.take().expect("Must be set in the constructor"));
+                }
+                2 => {
+                    let request = &self.request_msg;
+                    if let Some(association) = jet_associations.get_mut(&request.association) {
+                        if association.version() == JET_VERSION_V2 {
+                            if let Some(candidate) = association.get_candidate_mut(request.candidate) {
+                                candidate.set_transport(self.transport.take().expect("Must be set in the constructor"));
+                            }
+                        }
+                    }
+                }
+                _ => {
+                    // No jet message exist with version different than 1 or 2
+                    unreachable!()
+                }
+            };
+
+            if let Some(remove_association_future) = self.remove_association_future.take() {
+                self.executor_handle.spawn(remove_association_future);
```

Comment:
And if everything is ok, then do spawn for remove_association_future (as before)